### PR TITLE
feat: add retry stream client

### PR DESCRIPTION
**Pull request by bakkelim**
Retry stream client for when streams go down.

Link to issue: https://indykite.atlassian.net/browse/ENG-950

### CODE REVIEW

File: ingest/ingest_retry_client.go [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+// sendRecordWithRetry creates a new stream and tries sends a record. Will retry based on retry policy.
+func (c *RetryClient) sendRecordWithRetry(record *ingestpb.Record) error {
+	backoffTime, _ := time.ParseDuration(c.retryPolicy.InitialBackoff)
+	var err error
+	for i := 1; i <= c.retryPolicy.MaxAttempts; i++ {
+		log.Printf("attempting to reconnect (%d/%d) in %s...", i, c.retryPolicy.MaxAttempts, backoffTime.String())
+		time.Sleep(backoffTime)
+		backoffTime *= time.Duration(c.retryPolicy.BackoffMultiplier)
+		ctx, _ := context.WithTimeout(context.Background(), c.retryPolicy.CtxTimeout) //nolint
```
Comment (bakkelim) [inline ingest/ingest_retry_client.go:129]: I'm unsure about this one, linter complained if I did not use and call the cancel func.

Comment (bakkelim) [reply ingest/ingest_retry_client.go:129]: Yes, but the linter still complains because the cancel func is in a for loop. I've also tried to put the declaration of the func outside the for loop, but still lint issues. 🤷‍♂️

File: examples/ingest/cmd/root.go [hunk 2]
```diff
@@ -110,6 +112,19 @@ func initConfig() {
...
 	}
+	retryClient, err = ingest.NewRetryClient(context.Background(),
+		&ingest.RetryPolicy{
+			MaxAttempts:       4,
+			InitialBackoff:    "1s",
+			BackoffMultiplier: 2,
+			CtxTimeout:        120 * time.Second,
+		},
+		grpc.WithCredentialsLoader(apicfg.DefaultEnvironmentLoader),
+		grpc.WithRetryOptions(retry.Disable()),
```
Comment (maaland) [inline examples/ingest/cmd/root.go:123]: Maybe we should hardcode the `grpc.WithRetryOptions(retry.Disable())` in the retry client? Looks a bit weird to pass `retry.Disable()` to a retry client 😅

File: ingest/ingest_retry_client.go [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+	"google.golang.org/grpc/codes"
+
+	"github.com/indykite/indykite-sdk-go/errors"
+	ingestpb "github.com/indykite/indykite-sdk-go/gen/indykite/ingest/v1beta2"
+	api "github.com/indykite/indykite-sdk-go/grpc"
+)
+
+type RetryPolicy struct {
+	InitialBackoff    string
```
Comment (maaland) [inline ingest/ingest_retry_client.go:31]: maybe some documentation on what values the `InitialBackoff` supports?

File: ingest/ingest_retry_client.go [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+		retryPolicy: retryPolicy,
+	}
+
+	if _, err = time.ParseDuration(c.retryPolicy.InitialBackoff); err != nil {
+		return nil, errors.New(codes.FailedPrecondition, fmt.Sprintf("invalid backoff duration: %v", err))
+	}
+	return c, nil
+}
+
+// NewTestRetryClient creates a new Ingest API gRPC Client with retry functionality.
```
Comment (maaland) [inline ingest/ingest_retry_client.go:60]:
```suggestion
// NewTestRetryClient creates a new test Ingest API gRPC Client with retry functionality.
```

File: ingest/ingest_retry_client.go [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+// SendRecord sends a record on the opened stream and retries if it fails.
+func (c *RetryClient) SendRecord(record *ingestpb.Record) error {
+	if c.stream == nil {
+		return errors.New(codes.FailedPrecondition, "a stream must be opened first")
+	}
+
+	err := c.Client.SendRecord(record)
+	if err == nil {
+		return err
```
Comment (maaland) [inline ingest/ingest_retry_client.go:88]: Not sure what the best practice is here, but at first glance it's a little confusing to read `return err` when the `err` is in fact `nil` in this case. Would it be better to `return nil` here?

File: ingest/ingest_retry_client.go [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+	}
+	return c.sendRecordWithRetry(record)
+}
+
+// ReceiveResponse returns the next response available on the stream.
```
Comment (maaland) [inline ingest/ingest_retry_client.go:93]: Add a comment about how it waits/retries

File: ingest/ingest_retry_client.go [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+
+	resp, err := c.Client.ReceiveResponse()
+	if err == nil {
+		return resp, err
+	}
+	return c.receiveResponseWhenReconnected()
+}
+
+// receiveResponseWhenReconnected returns try to read the next response available on the stream
+// or return error if unable to reconnect.
```
Comment (maaland) [inline ingest/ingest_retry_client.go:107]: this sentence doesn't really make sense, rephrase? 😅

File: ingest/ingest_retry_client.go (hunk 5 above)
Comment (maaland) [reply ingest/ingest_retry_client.go:88]: The same applies to a few other occurrences.

File: ingest/ingest_retry_client.go [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+		ctx, _ := context.WithTimeout(context.Background(), c.retryPolicy.CtxTimeout) //nolint
+		err = c.OpenStreamClient(ctx)
+		if err != nil {
+			continue
+		}
+		time.Sleep(500 * time.Millisecond)
+		err = c.Client.SendRecord(record)
+		if err == nil {
+			log.Printf("restablished connection to server")
+			return err
```
Comment (maaland) [inline ingest/ingest_retry_client.go:138]: What happens if the sleep here is too short for whatever reason, to the `SendRecord` fails but shortly after the stream is indeed opened. Then, we loop again and try to open a stream again. Is that okay?

**Review by maaland — COMMENTED** [review]
Looks good, just a few comments/questions!

File: ingest/ingest_retry_client.go (hunk 3 above)
Comment (arxeiss) [reply ingest/ingest_retry_client.go:31]: Why it is not of type `time.Duration` already, when you parse it to time.Duration anyway?

Comment (bakkelim) [reply ingest/ingest_retry_client.go:31]: I based it on this policy suggested by grpc team. So want to keep as similar as possible. https://github.com/grpc/proposal/blob/master/A6-client-retries.md#retry-policy-capabilities

File: examples/ingest/cmd/root.go (hunk 2 above)
Comment (bakkelim) [reply examples/ingest/cmd/root.go:123]: fixed

File: ingest/ingest_retry_client.go (hunk 3 above)
Comment (arxeiss) [reply ingest/ingest_retry_client.go:31]: They say, it must follow JSON structure of underlying Proto message. You have no underlying Proto nor JSON. So I would go with time.Duration directly.

File: ingest/ingest_retry_client.go (hunk 5 above)
Comment (bakkelim) [reply ingest/ingest_retry_client.go:88]: fixed

File: ingest/ingest_retry_client.go (hunk 6 above)
Comment (bakkelim) [reply ingest/ingest_retry_client.go:93]: fixed

File: ingest/ingest_retry_client.go (hunk 7 above)
Comment (bakkelim) [reply ingest/ingest_retry_client.go:107]: fixed

File: ingest/ingest_retry_client.go (hunk 8 above)
Comment (bakkelim) [reply ingest/ingest_retry_client.go:138]: I don't know. I guess we need to test and see if this in an issue

File: examples/ingest/cmd/stream_retry.go [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+			err := retryClient.SendRecord(record)
+			if err != nil {
+				log.Fatalf("failed to send record %v", err)
+			}
+			resp, err := retryClient.ReceiveResponse()
+			if err != nil {
+				log.Fatalf("failed to receive response %v", err)
+			}
+			log.Println(jsonp.Format(resp))
+		}
```
[the remark this answers is not in the document]
Comment (bakkelim) [reply examples/ingest/cmd/stream_retry.go:118]: You could, and it's up to the developer implementing this. This I just an example and it's easiest to show with synchronous call

File: ingest/ingest_retry_client.go [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+// sendRecordWithRetry creates a new stream and tries sends a record. Will retry based on retry policy.
+func (c *RetryClient) sendRecordWithRetry(record *ingestpb.Record) error {
+	backoffTime, _ := time.ParseDuration(c.retryPolicy.InitialBackoff)
+	var err error
+	for i := 1; i <= c.retryPolicy.MaxAttempts; i++ {
+		log.Printf("attempting to reconnect (%d/%d) in %s...", i, c.retryPolicy.MaxAttempts, backoffTime.String())
+		time.Sleep(backoffTime)
+		backoffTime *= time.Duration(c.retryPolicy.BackoffMultiplier)
+		ctx, _ := context.WithTimeout(context.Background(), c.retryPolicy.CtxTimeout) //nolint
+		err = c.OpenStreamClient(ctx)
```
[the remark this answers is not in the document]
Comment (bakkelim) [reply ingest/ingest_retry_client.go:130]: Why? Only send is calling this

File: ingest/ingest_retry_client.go (hunk 3 above)
Comment (bakkelim) [reply ingest/ingest_retry_client.go:32]: Agree to disagree 😄 
I'll keep it this way as it's only used when having to reconnect and should not affect perfomance

File: ingest/ingest_retry_client.go (hunk 1 above)
Comment (jakubtomany) [reply ingest/ingest_retry_client.go]: Here is nice explanation: https://stackoverflow.com/a/45620423

TLDR: wrap the content of the loop in a function (either anonymous or named one)

File: ingest/ingest_retry_client.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (arxeiss) [reply ingest/ingest_retry_client.go:134]: Was looking at wrong place, sorry

File: examples/ingest/cmd/root.go (hunk 2 above)
Comment (maaland) [reply examples/ingest/cmd/root.go]: Good question, and the answer is "kinda". The built-in gRPC retry functionality does not work for streams, so if you call the stream function of a client with the retry options enabled, you will get a gRPC error. But it would be useful for the single ingest function (non-stream), which the client also has.

Comment (bakkelim) [reply examples/ingest/cmd/root.go]: You cannot allow retry since the client contains stream methods.
image

Comment (maaland) [reply examples/ingest/cmd/root.go]: Yup, so we would need to find a way to enable that only in the single `IngestRecord()` function, but not in the stream. Not important for now, and out of scope for this PR

File: ingest/ingest_retry_client.go (hunk 1 above)
Comment (bakkelim) [reply ingest/ingest_retry_client.go]: Context issue solve!

**Review by jakubtomany — APPROVED** [review]
LGTM

File: ingest/ingest_retry_client.go (hunk 4 above)
Comment (maaland) [reply ingest/ingest_retry_client.go:62]: did you?

**Review by maaland — APPROVED** [review]
Great work!

File: examples/ingest/cmd/root.go [hunk 11]
```diff
@@ -110,6 +111,17 @@ func initConfig() {
 	if err != nil {
 		er(fmt.Sprintf("failed to create IndyKite Ingest Client: %v", err))
 	}
+	retryClient, err = ingest.NewRetryClient(context.Background(),
+		&ingest.RetryPolicy{
+			MaxAttempts:       4,
+			InitialBackoff:    "1s", // proto duration format according to spec: https://protobuf.dev/programming-guides/proto3/#json
```
Comment (arxeiss) [inline examples/ingest/cmd/root.go:117]: Really, I don't like this:

1. You are referring to JSON format of Proto message. But this `RetryPolicy` is not Proto message. And it is not possible to Marshall or Unmarshal to JSON.
2. You are parsing it again and again. If client fails, you cannot just grab backoff, you need to parse this string to duration again. Requires extra CPU time for nothing.
3. You are referring to Proto spec, which requires format `1s` or `1.000001s`. But if I pass `30ms` which is not valid Proto duration, it will still work. Because time.Duration supports that format.
4. Ad 1&3. If you would implement JSON parsing. Time duration could generate time like `30ms` which is again not valid according to spec you pointed out.

Comment (bakkelim) [reply examples/ingest/cmd/root.go:117]: fine, I'll change it

File: examples/ingest/cmd/root.go [hunk 12]
```diff
@@ -110,6 +112,17 @@ func initConfig() {
 	if err != nil {
 		er(fmt.Sprintf("failed to create IndyKite Ingest Client: %v", err))
 	}
+	retryClient, err = ingest.NewRetryClient(context.Background(),
+		&ingest.RetryPolicy{
+			MaxAttempts:       4,
+			InitialBackoff:    1 * time.Second, // proto duration format according to spec: https://protobuf.dev/programming-guides/proto3/#json
```
Comment (arxeiss) [inline examples/ingest/cmd/root.go:118]: just remove the comment :)

**Review by arxeiss — APPROVED** [review]

File: ingest/ingest_retry_client.go (hunk 4 above)
Comment (maaland) [reply ingest/ingest_retry_client.go:57]: you updated the wrong comment 😄
